### PR TITLE
Consolidate CFString conversions

### PR DESCRIPTION
**Pull request by ePirat**
This commitset aims to consolidate conversions of CoreFoundations `CFStringRef` to C `char *` or OBS‘ `dstr`.

### CODE REVIEW

File: libobs/util/apple/cfstring-utils.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#pragma once
+
+#include "../c99defs.h"
+#include "../dstr.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+EXPORT char* cfstr_copy_cstr(CFStringRef cfstr, CFStringEncoding cfstr_enc);
```
Comment (RytoEX) [inline libobs/util/apple/cfstring-utils.h]: Style:  Place the asterisk next to the function name.

File: libobs/util/apple/cfstring-utils.h [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+EXPORT bool cfstr_copy_dstr(CFStringRef cfstr, CFStringEncoding cfstr_enc,
+    struct dstr *str);
```
Comment (RytoEX) [inline libobs/util/apple/cfstring-utils.h]: Style:  Use two tabs to indent line continuations.  Do not use spaces for indents.

File: libobs/util/platform-cocoa.m [hunk 3]
```diff
@@ -417,3 +420,89 @@ uint64_t os_get_proc_virtual_size(void)
 		return 0;
 	return taskinfo.virtual_size;
 }
+
+/* Obtains a copy of the contents of a CFString in specified encoding.
+ * Returns char* (must be bfree'd by caller) or NULL on failure.
+ */
+char* cfstr_copy_cstr(CFStringRef cfstring, CFStringEncoding cfstring_encoding)
```
Comment (RytoEX) [inline libobs/util/platform-cocoa.m]: Style: Place the asterisk next to the function name.

File: libobs/util/platform-cocoa.m [hunk 4]
```diff
@@ -417,3 +420,89 @@ uint64_t os_get_proc_virtual_size(void)
...
+{
+	if (!cfstring) return NULL;
```
Comment (RytoEX) [inline libobs/util/platform-cocoa.m]: Style:  Don't include multiple statements on a single line.  Place the `return` on its own line.

File: libobs/util/platform-cocoa.m [hunk 5]
```diff
@@ -417,3 +420,89 @@ uint64_t os_get_proc_virtual_size(void)
...
+
+	// Try the quick way to obtain the buffer
+	const char *tmp_buffer = CFStringGetCStringPtr(cfstring, cfstring_encoding);
```
Comment (RytoEX) [inline libobs/util/platform-cocoa.m]: Style:  Line exceeds 80 columns.  Line-break this line and indent accordingly.

File: libobs/util/platform-cocoa.m [hunk 6]
```diff
@@ -417,3 +420,89 @@ uint64_t os_get_proc_virtual_size(void)
...
+
+	if (tmp_buffer != NULL) {
```
Comment (RytoEX) [inline libobs/util/platform-cocoa.m]: Style:  This `if` statement does not need curly braces.  Remove them.

File: libobs/util/platform-cocoa.m [hunk 7]
```diff
@@ -417,3 +420,89 @@ uint64_t os_get_proc_virtual_size(void)
...
+		return bstrdup(tmp_buffer);
+	}
+
+	// The quick way did not work, try the more expensive one
+	CFIndex length = CFStringGetLength(cfstring);
+	CFIndex max_size =
+		CFStringGetMaximumSizeForEncoding(length, cfstring_encoding);
+
+	// If result would exceed LONG_MAX, kCFNotFound is returned
+	if (max_size == kCFNotFound) {
```
Comment (RytoEX) [inline libobs/util/platform-cocoa.m]: Style:  This `if` statement does not need curly braces.  Remove them.

File: libobs/util/platform-cocoa.m [hunk 8]
```diff
@@ -417,3 +420,89 @@ uint64_t os_get_proc_virtual_size(void)
...
+		return NULL;
+	}
+
+	// Account for the null terminator
+	max_size++;
+
+	char *buffer = bmalloc(max_size);
+
+	if (buffer == NULL) {
```
Comment (RytoEX) [inline libobs/util/platform-cocoa.m]: Style:  This `if` statement does not need curly braces.  Remove them.

File: libobs/util/platform-cocoa.m [hunk 9]
```diff
@@ -417,3 +420,89 @@ uint64_t os_get_proc_virtual_size(void)
...
+	}
+	return buffer;
+}
+
+/* Copies the contents of a CFString in specified encoding to a given dstr.
+ * Returns true on success or false on failure.
+ * In case of failure, the dstr capacity but not size is changed.
+ */
+bool cfstr_copy_dstr(CFStringRef cfstring,
+	CFStringEncoding cfstring_encoding, struct dstr *str)
```
Comment (RytoEX) [inline libobs/util/platform-cocoa.m]: Style: Use two tabs to indent line continuations.

You could have inserted the line-break after the second parameter, which is what I'd prefer, but this is probably acceptable.

File: libobs/util/platform-cocoa.m [hunk 10]
```diff
@@ -417,3 +420,89 @@ uint64_t os_get_proc_virtual_size(void)
...
+{
+	if (!cfstring) return false;
```
Comment (RytoEX) [inline libobs/util/platform-cocoa.m]: Style:  Don't include multiple statements on a single line.  Place the `return` on its own line.

File: libobs/util/platform-cocoa.m [hunk 11]
```diff
@@ -417,3 +420,89 @@ uint64_t os_get_proc_virtual_size(void)
...
+
+	// Try the quick way to obtain the buffer
+	const char *tmp_buffer = CFStringGetCStringPtr(cfstring, cfstring_encoding);
```
Comment (RytoEX) [inline libobs/util/platform-cocoa.m]: Style: Line exceeds 80 columns. Line-break this line and indent accordingly.

File: libobs/util/platform-cocoa.m [hunk 12]
```diff
@@ -417,3 +420,89 @@ uint64_t os_get_proc_virtual_size(void)
...
+	max_size++;
+
+	char *buffer = bmalloc(max_size);
+
+	if (buffer == NULL) {
+		return NULL;
+	}
+
+	// Copy CFString in requested encoding to buffer
+	Boolean success =
```
Comment (RytoEX) [inline libobs/util/platform-cocoa.m]: I think "boolean" is usually lowercased.  Unless this is special somehow?

File: libobs/util/platform-cocoa.m [hunk 13]
```diff
@@ -417,3 +420,89 @@ uint64_t os_get_proc_virtual_size(void)
...
+		return NULL;
+	}
+
+	// Account for the null terminator
+	max_size++;
+
+	dstr_ensure_capacity(str, max_size);
+
+	// Copy CFString in requested encoding to dstr buffer
+	Boolean success = CFStringGetCString(
```
Comment (RytoEX) [inline libobs/util/platform-cocoa.m]: I think "boolean" is usually lowercased.  Unless this is special somehow?

File: libobs/util/platform-cocoa.m [hunk 14]
```diff
@@ -417,3 +420,89 @@ uint64_t os_get_proc_virtual_size(void)
...
+		return true;
+	}
+
+	// The quick way did not work, try the more expensive one
+	CFIndex length = CFStringGetLength(cfstring);
+	CFIndex max_size =
+		CFStringGetMaximumSizeForEncoding(length, cfstring_encoding);
+
+	// If result would exceed LONG_MAX, kCFNotFound is returned
+	if (max_size == kCFNotFound) {
```
Comment (RytoEX) [inline libobs/util/platform-cocoa.m]: Style:  This `if` statement does not need curly braces.  Remove them.

File: libobs/util/platform-cocoa.m (hunk 12 above)
Comment (ePirat) [reply libobs/util/platform-cocoa.m]: `Boolean` is a special macOS thing, afaik. It is defined in `MacTypes.h` and written with an uppercase `B`.

File: libobs/util/platform-cocoa.m [hunk 15]
```diff
@@ -417,3 +420,91 @@ uint64_t os_get_proc_virtual_size(void)
...
+ * Returns char* (must be bfree'd by caller) or NULL on failure.
+ */
+char *cfstr_copy_cstr(CFStringRef cfstring, CFStringEncoding cfstring_encoding)
+{
+	if (!cfstring)
+		return NULL;
+
+	// Try the quick way to obtain the buffer
+	const char *tmp_buffer = CFStringGetCStringPtr(cfstring,
+		cfstring_encoding);
```
Comment (RytoEX) [inline libobs/util/platform-cocoa.m]: Style:  Indent a line continuation with two tabs from the starting indent.

File: libobs/util/platform-cocoa.m [hunk 16]
```diff
@@ -417,3 +420,91 @@ uint64_t os_get_proc_virtual_size(void)
...
+
+	if (tmp_buffer != NULL) {
```
Comment (RytoEX) [inline libobs/util/platform-cocoa.m]: Style:  This `if` statement does not need curly braces.  Remove them.

File: libobs/util/platform-cocoa.m [hunk 17]
```diff
@@ -417,3 +420,91 @@ uint64_t os_get_proc_virtual_size(void)
...
+ */
+bool cfstr_copy_dstr(CFStringRef cfstring,
+	CFStringEncoding cfstring_encoding, struct dstr *str)
+{
+	if (!cfstring)
+		return false;
+
+	// Try the quick way to obtain the buffer
+	const char *tmp_buffer = CFStringGetCStringPtr(cfstring,
+		cfstring_encoding);
```
Comment (RytoEX) [inline libobs/util/platform-cocoa.m]: Style:  Indent a line continuation with two tabs from the starting indent.

**Pull request merged**